openai/client.py

```diff
@@ -0,0 +1,401 @@
+                "https://cognitiveservices.azure.com/.default"
+            )
+
+        return self._cached_token.token
+
+
+class ApiKeyAuth:
+    """Authentication using an API key"""
+
+    def __repr__(self):
+        return f"ApiKeyAuth(api_key=<redacted>)"
+    
+    def __init__(self, key: str = ""):
+        """Create a new ApiKeyAuth instance. If no key is passed, it will use ~openai.api_key"""
+        self.key = key or openai.api_key
+
+    def get_token(self) -> str:
+        """Get the API key"""
+        return self.key
+
+
+Backends = Literal["azure", "openai", ""]
+
+
+class OpenAIClient:
+    def __init__(
+        self,
+        *,
+        api_base: str = "",
+        auth: Union[str, ApiKeyAuth, AzureTokenAuth] = "",
+        api_version: str = "",
+        backend: Backends = "",
+        organization: str = "",
+    ):
+        """Create a new OpenAI client.
+           
+       :keyword str api_base: The base URL for the API. If not specified, based on ~opeanai.api_base
+       :keyword auth: The authentication method or key to use. If the string value "azuredefault" is passed, it will use ~azure.identity.DefaultAzureCredential
+       :paramtype auth: str or ~openai.client.ApiKeyAuth or ~openai.client.AzureTokenAuth
+       :keyword str api_version: The API version to use. If not specified, based on ~openai.api_version or ~openai.client.LATEST_AZURE_API_VERSION.
+       :keyword str backend: One of 'azure' or 'openai'. If not specified, inferred from the auth method or ~openai.api_type
+        """
+
+        #
+        # This code is a bit messy, but it's because we want to hide the messiness from the caller.
+        #
+        if auth == "azuredefault":
+            self.auth = AzureTokenAuth()
+        elif isinstance(auth, str):
+            self.auth = ApiKeyAuth(auth or openai.api_key)
+        else:
+            self.auth = auth
+
+        # Pick up api type from parameter or environment
+        self.backend = backend or (
+            "azure" if openai.api_type in ("azure", "azure_ad", "azuread") or isinstance(auth, AzureTokenAuth) else "openai"
+        )
+
+        if self.backend == "azure":
+            self.api_version = (
+                api_version or openai.api_version or LATEST_AZURE_API_VERSION
+            )
+            if isinstance(self.auth, AzureTokenAuth):
+                self.api_type = "azure_ad"
+            else:
+                self.api_type = "azure"
+        elif self.backend == "openai":
+            self.api_type = "open_ai"
+            self.api_version = api_version or openai.api_version
+        else:
+            raise ValueError(
+                f'Unknown `backend` {backend} - expected one of "azure" or "openai"'
+            )
+
+        self.api_base = api_base or openai.api_base
+        self.organization = organization or openai.organization
+        if self.backend == 'azure' and self.api_base == "https://api.openai.com/v1":
```

Comment:
what do you think about only comparing the api_base string up until the `v` or excluding the version altogether? e.g. `self.api_base.startswith("https://api.openai.com/v")`? I'm not sure how often the version is bumped, but it could possibly save us a re-release...